### PR TITLE
Add support for flow type spread

### PR DESCRIPTION
**Pull request by conartist6**
| Q                 | A
| ----------------- | ---
| Bug fix?          | no
| Breaking change?  | no
| New feature?      | yes
| Deprecations?     | no
| Spec compliancy?  | yes (matches flow parser behavior)
| Tests added/pass? | yes
| Fixed tickets     | https://github.com/babel/babylon/issues/416
| License           | MIT

This change is to support Flow's brand new type spread behavior. make test-flow tests captured the failing behavior, the 5 tests under the heading `### Object type spread ###` now pass.

### CODE REVIEW

File: src/plugins/flow.js [hunk 1]
```diff
@@ -450,24 +450,34 @@ pp.flowParseObjectType = function (allowStatic, allowExact) {
...
-        // This is a method property
-        if (variance) {
-          this.unexpected(variance.start);
+      if (this.match(tt.ellipsis)) {
+        if (!allowSpread) {
+          this.unexpected();
         }
-        nodeStart.properties.push(this.flowParseObjectTypeMethod(startPos, startLoc, isStatic, propertyKey));
+        this.expect(tt.ellipsis);
+        node.argument = this.flowParsePrimaryType();
```
Comment (gabelevi) [inline src/plugins/flow.js]: Should this be `this.flowParseType()`? That seems to be what the [Flow parser is doing](https://github.com/facebook/flow/blob/b33ce6ea80a426e8e971ab08a5bc3a107fd9ab33/src/parser/type_parser.ml#L519)

File: src/plugins/flow.js [hunk 2]
```diff
@@ -450,24 +450,34 @@ pp.flowParseObjectType = function (allowStatic, allowExact) {
...
       nodeStart.callProperties.push(this.flowParseObjectTypeCallProperty(node, isStatic));
     } else {
-      propertyKey = this.flowParseObjectPropertyKey();
-      if (this.isRelational("<") || this.match(tt.parenL)) {
-        // This is a method property
-        if (variance) {
-          this.unexpected(variance.start);
+      if (this.match(tt.ellipsis)) {
+        if (!allowSpread) {
+          this.unexpected();
```
Comment (danez) [inline src/plugins/flow.js]: Can we add a test for this here in a case where spread is not allowed?

File: test/fixtures/flow/type-annotations/136/actual.js [hunk 3]
```diff
@@ -0,0 +1,4 @@
+type A = {
+	p: {},
+	...{},
+};
```
Comment (danez) [inline test/fixtures/flow/type-annotations/136/actual.js]: This should not parse I think. Probably related to what @gabelevi wrote.

**Comment by danez** [comment]
Thanks for doing this. 👍

File: test/fixtures/flow/type-annotations/135/actual.js [hunk 4]
```diff
@@ -0,0 +1,3 @@
+type A = {
+	...any,
```
Comment (danez) [inline test/fixtures/flow/type-annotations/135/actual.js]: any is also not allowed here it seems, flow is throwing an error. Seems to also happen for all primitive types: `number, string, boolean, any, null, ...`

Comment (samwgoldman) [inline test/fixtures/flow/type-annotations/135/actual.js]: This should parse. How are you testing?

File: test/fixtures/flow/type-annotations/136/actual.js (hunk 3 above)
Comment (samwgoldman) [inline test/fixtures/flow/type-annotations/136/actual.js]: This should parse. How are you testing? Gabe's comment is that parsePrimaryType is actually *too* restrictive. We also parse non-primary types in this position, like unions. E.g., `{...A|B}`.

Comment (danez) [inline test/fixtures/flow/type-annotations/136/actual.js]: Oh okay sorry. I see astexplorer uses flow 0.40 which supports it spreading only for Identifiers?

Comment (samwgoldman) [inline test/fixtures/flow/type-annotations/136/actual.js]: Right, we "reserved" the syntax in 0.40, but 0.42 extends support to all type syntax and adds support.

Comment (gabelevi) [inline test/fixtures/flow/type-annotations/136/actual.js]: Yeah, PR for astexplorer hasn't been merged yet https://github.com/fkling/astexplorer/pull/208

Comment (gabelevi) [inline test/fixtures/flow/type-annotations/136/actual.js]: But https://flowtype.org/try/ should work and it has an AST tab

File: test/fixtures/flow/type-annotations/135/actual.js (hunk 4 above)
Comment (danez) [inline test/fixtures/flow/type-annotations/135/actual.js]: Okay was testing on astexplorer, which is a bad idea. Primitives, besides `any`  fail though if I see correctly. https://flowtype.org/try/#0C4TwDgpgBAglC8UDeAoKBIAdNgRgezwBsIBDAOwBo0BfAbiA

Comment (samwgoldman) [inline test/fixtures/flow/type-annotations/135/actual.js]: Yeah, no worries. Per your initial comment, the typing support is incomplete, but eventually we will also support spreading `string` (implicitly promotes to `String` instance, which has no own properties).

File: test/fixtures/flow/type-annotations/136/actual.js (hunk 3 above)
Comment (hzoo) [inline test/fixtures/flow/type-annotations/136/actual.js]: We should see if we can setup greenkeeper or a bot to update the package + yarn @fkling

Comment (danez) [inline test/fixtures/flow/type-annotations/136/actual.js]: I think felix is working on this. https://github.com/fkling/astexplorer/pull/197

**Comment by samwgoldman** [comment]
If I remember correctly, we still need to put up a PR to babel/babel with changes to babel-types, babel-generator, and babel-transform-flow-strip-types, correct?

**Comment by conartist6** [comment]
I'll look into that next.

**Comment by hzoo** [comment]
@samwgoldman wrote up a simple wiki page https://github.com/babel/babel/wiki/Adding-a-new-Proposal-to-Babel

**Comment by hzoo** [comment]
We moved the previous 7.0 branch to be `master` now, and `master` is now 6.x

**Comment by conartist6** [comment]
You're saying that I need to alter the PR? Which branch should I be requesting to merge to?

**Comment by hzoo** [comment]
No changes, just informing. We just need to update Babel

**Comment by conartist6** [comment]
Pull request for babel is open: https://github.com/babel/babel/pull/5525
Added a few other minor changes to this PR.

**Comment by danez** [comment]
Interesting that flow parses the testcase you added in your last commit. I added an issue in the flow repo.
https://github.com/facebook/flow/issues/3566

**Comment by zertosh** [comment]
Any chance this can be backported to Babylon 6?

**Comment by milesj** [comment]
Has this been released in `babel-eslint` yet? I'm still getting failures when spreading.

**Pull request merged**